File: database/models/paper.go
package models

import "github.com/pgvector/pgvector-go"

type Paper struct {
	ID        uint64          `db:"id" json:"id"`
	Title     string          `db:"title" json:"title"`
	Abstract  string          `db:"abstract" json:"abstract"`
	DOI       string          `db:"doi" json:"doi"`
	Embedding pgvector.Vector `db:"embedding" json:"embedding"`
}

File: app/lib/models/paper.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'paper.freezed.dart';
part 'paper.g.dart';

@freezed
sealed class Paper with _$Paper {
  factory Paper({
    required String doi,
    required String title,
    required String abstract,
  }) = _Paper;

  factory Paper.fromJson(Map<String, dynamic> json) => _$PaperFromJson(json);
}

@freezed
sealed class FindPaperQuery with _$FindPaperQuery {
  factory FindPaperQuery({required String query}) = _FindPaperQuery;

  factory FindPaperQuery.fromJson(Map<String, dynamic> json) =>
      _$FindPaperQueryFromJson(json);
}
